docs/rfcs/2025-08-11-channels-forwarding.md

```diff
@@ -0,0 +1,143 @@
+
+Schema draft:
+
+```sql
+ALTER TABLE messages ADD COLUMN chat_message_json TEXT;
+
+CREATE TABLE forwarding_instructions (
+  forwarding_instruction_id INTEGER PRIMARY KEY,
+  group_id INTEGER NOT NULL REFERENCES groups ON DELETE CASCADE,
+  forward_scope TEXT NOT NULL, -- save as JSON/text? add field for scope group member id for GFSMemberSupport scope?
+  sending_group_member_id INTEGER NOT NULL REFERENCES group_members(group_member_id) ON DELETE CASCADE, -- or sending_member_id BLOB without fkey
+  broker_ts TEXT NOT NULL,
+  message_ids TEXT NOT NULL, -- comma separated list; normalize via many-to-many table? doesn't seem necessary
+  failed INTEGER DEFAULT 0, -- for worker marking forwarding instruction as failed, to be able to proceed to next one
+  created_at TEXT NOT NULL DEFAULT (datetime('now')),
+  updated_at TEXT NOT NULL DEFAULT (datetime('now'))
+);
+```
+
+Forwarding worker will read forwarding instructions sorted by created_at, retrieve chat messages corresponding to message_ids, retrieve group members based on group_id and forward_scope and send batched forward events to them.
+
+How to split forwarding to large number of group members into smaller batches?
+
+One option is to determine forwarding lists at the moment of creating forwarding instruction and persist them, possibly already split into predefined batches. However, this unnecessarily excludes newly joined members (those that joined between creating instruction and worker picking it up for forwarding operation).
+
+Another approach would be to split retrieved members into batches at time of processing instruction, and persist group member ids that instruction was processed for after sending one batch at a time. Group member retrieval should then filter out already processed members in case of operation failure leading to repeat processing.
+
+```sql
+CREATE TABLE forwarding_instructions_members (
+  forwarding_instruction_member_id INTEGER PRIMARY KEY,
+  forwarding_instruction_id INTEGER NOT NULL REFERENCES forwarding_instructions ON DELETE CASCADE,
+  group_member_id INTEGER NOT NULL REFERENCES group_members ON DELETE CASCADE
+);
+```
+
+Overall forwarding worker algorithm would then be:
+1. Retrieve next forwarding instruction.
+    - Retrieve chat messages corresponding to message_ids, build forward events (XGrpMsgForward).
+    - Retrieve member list based on group_id and forward_scope, filtering out already processed members via forwarding_instructions_members.
+2. Split member list into smaller batches. TBC how to choose optimal batch size.
+3. For each batch of members:
+    1. Group send (sendGroupMessages_) batch of XGrpMsgForward events.
+    2. Persist processed member ids to forwarding_instructions_members.
+        - Persist after save is ok because receiving client can deduplicate in case chat relay fails in-between.
+    3. Possibly small delay to avoid overloading database.
```

Comment:
there is no need for it, we have two separate pools already, these jobs would use low priority pool